lib/vdsm/virt/thinp.py

```diff
@@ -333,36 +362,18 @@ def _should_extend_volume(self, drive, volumeID, block_info):
             # next lvm extent.
             return False
 
-        if (block_info.allocation == 0 and
-                drive.threshold_state == storage.BLOCK_THRESHOLD.EXCEEDED):
-            # We get allocation == 0:
-            # - Before the guest write to the disk.
-            # - Older libvirt versions did not report allocation during
-            #   backup, see https://bugzilla.redhat.com/2015281.
-            # If we got a threshold event, we can safely assume that the guest
-            # wrote to the drive, and we need to extend it.
-            self._log.warning(
-                "No allocation info for drive %s, but block threshold was "
-                "exceeded - assuming that drive needs extension",
-                drive.name)
+        # If drive is marked for extension, extend it regardless of allocation.
```

Comment:
This is a good thing to do but hopefully it cannot happen that drive.threshold_state keeps storage.BLOCK_THRESHOLD.EXCEEDED under some edge case and the drive gets exceeded again and again.  I’m not familiar with the flows and corresponding protections, I assume they are there, just pointing out that this may be worth of double checking.

Reply:
While a drive is mark as EXCEEDED, it is possible that extend timeout has expired, and the periodic monitor will try to extend again. If the drive was not refreshed yet, the next extend will be a duplicate extend to the same size. This is already handled by the mailbox and at lvm level, so it is harmless.

When we refresh a drive successfully, we set new block threshold and set the drive state to SET. At this point thread will not try to extend the drive.

If we failed to set a new block threshold, the drive is marked as UNSET, and the next monitor attempt will retry to set a new block threshold.

We have a small window - after the drive is refreshed, but before we mark it as SET or UNSET. If at this time another thread will try to monitor, it can try to extend again when there is no need to extend.

This race is not new, but it can cause double extend now that we don't check the allocation. I think to lock the drive or the monitor before we refresh it until we finish to set a new block threshold to close this window.

Reply:
OK, it doesn’t look as a risk with big consequences (like unnecessarily extending many times) and can be addressed later.  Keeping unresolved for now to not forget about it.